### PR TITLE
Implement discounts management in Order View page

### PR DESCRIPTION
**Pull request by sarjon**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR implements discounts management in new Order view page.
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/15826
| How to test?  | Please indicate how to best verify that this PR is correct.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/15901)

### CODE REVIEW

File: admin-dev/themes/new-theme/js/pages/order/view.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    $form.find(OrderViewPageMap.addCartRuleApplyOnAllInvoicesCheckbox).on('change', (event) => {
+      const isChecked = $(event.currentTarget).is(':checked');
+
+      $invoiceSelect.attr('disabled', isChecked);
+    });
+
+    $form.find(OrderViewPageMap.addCartRuleTypeSelect).on('change', (event) => {
+      const selectedCartRuleType = $(event.currentTarget).val();
+
+      if (selectedCartRuleType === 'amount') {
```
Comment (matks) [inline admin-dev/themes/new-theme/js/pages/order/view.js]:
```suggestion
      if (selectedCartRuleType === TYPE_AMOUNT) {
```

And at the top of the file, `const TYPE_AMOUNT = 'amount';`

Comment (matks) [reply admin-dev/themes/new-theme/js/pages/order/view.js]: This is like using constants in PHP 😉

File: admin-dev/themes/new-theme/js/pages/order/view.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        $valueHelp.removeClass('d-none');
+      } else {
+        $valueHelp.addClass('d-none');
+      }
+
+      if (selectedCartRuleType === 'free_shipping') {
```
Comment (matks) [inline admin-dev/themes/new-theme/js/pages/order/view.js]: Same

File: src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php [hunk 3]
```diff
@@ -79,8 +80,8 @@ public function handle(UpdateProductInOrderCommand $command)
         $product_quantity = $command->getQuantity();
 
         // @todo: use https://github.com/PrestaShop/decimal for price computations
-        $product_price_tax_incl = Tools::ps_round(Tools::getValue('product_price_tax_incl'), 2);
-        $product_price_tax_excl = Tools::ps_round(Tools::getValue('product_price_tax_excl'), 2);
+        $product_price_tax_incl = Tools::ps_round($command->getPriceTaxIncluded(), 2);
```
Comment (matks) [inline src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php:83]: @eternoendless Looks like there was a hardcoded "2 digits round" rule in legacy code for "Order View page". Shall we make it configurable, and use the shop digits number instead, or do we stay with this setting ?

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php [hunk 4]
```diff
@@ -237,7 +242,11 @@ public function viewAction(int $orderId): Response
         /** @var OrderForViewing $orderForViewing */
         $orderForViewing = $this->getQueryBus()->handle(new GetOrderForViewing($orderId));
 
-        $addOrderCartRuleForm = $this->createForm(AddOrderCartRuleType::class);
+        dump($orderForViewing);
```
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: Debug only 😄

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php [hunk 5]
```diff
@@ -246,13 +255,62 @@ public function viewAction(int $orderId): Response
...
+        ]);
+    }
+
+    /**
+     * @param int $orderId
+     * @param int $orderCartRuleId
+     *
+     * @return RedirectResponse
+     */
+    public function removeCartRuleAction(int $orderId, int $orderCartRuleId): RedirectResponse
```
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:292]: Dont forget security

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php [hunk 6]
```diff
@@ -246,13 +255,62 @@ public function viewAction(int $orderId): Response
...
+        );
+
+        $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+
+        return $this->redirectToRoute('admin_orders_view', [
+            'orderId' => $orderId,
+        ]);
+    }
+
+    public function updateProductAction(int $orderId, int $orderDetailId, Request $request): RedirectResponse
```
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:305]: Dont forget security

File: src/PrestaShopBundle/Form/Admin/Sell/Order/AddOrderProductType.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+namespace PrestaShopBundle\Form\Admin\Sell\Order;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\NumberType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+class AddOrderProductType extends AbstractType
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/Sell/Order/AddOrderProductType.php]:
```suggestion
class AddProductToOrderType extends AbstractType
```
Wdyt ?

File: src/PrestaShopBundle/Form/Admin/Sell/Order/UpdateOrderProductType.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Form\Admin\Sell\Order;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+class UpdateOrderProductType extends AbstractType
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/Sell/Order/UpdateOrderProductType.php]:
```suggestion
class UpdateProductFromAnOrderType extends AbstractType
```
Wdyt ?

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/orders/orders.yml [hunk 9]
```diff
@@ -79,3 +79,21 @@ admin_orders_add_payment:
     _controller: PrestaShopBundle:Admin/Sell/Order/Order:addPayment
   requirements:
     orderId: \d+
+
+admin_orders_update_product:
+  path: /{orderId}/products/{orderDetailId}
+  methods: [POST]
+  defaults:
+    _controller: PrestaShopBundle:Admin/Sell/Order/Order:updateProduct
```
Comment (matks) [inline src/PrestaShopBundle/Resources/config/routing/admin/sell/orders/orders.yml:115]: Legacy links ?
Maybe at the end of the order page project ?

**Review by matks — CHANGES_REQUESTED** [review]

File: admin-dev/themes/new-theme/js/pages/order/view.js (hunk 1 above)
Comment (sarjon) [reply admin-dev/themes/new-theme/js/pages/order/view.js]: Yeah, makes sense, for some reason I hesitated doing so. :confused:

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 4 above)
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: :man_facepalming: I have an excuse, it will be needed for next PRs anyways. :smile:

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 5 above)
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:292]: It will be done in last PR before enabling page, many PRs modify this file, so I want to have as little as possible conflicts.

File: src/PrestaShopBundle/Form/Admin/Sell/Order/AddOrderProductType.php (hunk 7 above)
Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Sell/Order/AddOrderProductType.php]: Sure. :+1: Don't forget I cannot merge these kind of suggestions, since it needs to be applied everywhere. :(

File: src/PrestaShopBundle/Form/Admin/Sell/Order/UpdateOrderProductType.php (hunk 8 above)
Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Sell/Order/UpdateOrderProductType.php]: How about `UpdateProductInOrderType`? :thinking:

File: src/PrestaShopBundle/Resources/config/routing/admin/sell/orders/orders.yml (hunk 9 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/routing/admin/sell/orders/orders.yml:115]: Yup, at the end of page migration, otherwise it will enable this page too early.

File: admin-dev/themes/new-theme/js/pages/order/view.js (hunk 1 above)
Comment (sarjon) [reply admin-dev/themes/new-theme/js/pages/order/view.js]: Done.

File: admin-dev/themes/new-theme/js/pages/order/view.js (hunk 2 above)
Comment (sarjon) [reply admin-dev/themes/new-theme/js/pages/order/view.js]: Done.

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 4 above)
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: Done.

File: src/PrestaShopBundle/Form/Admin/Sell/Order/AddOrderProductType.php (hunk 7 above)
Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Sell/Order/AddOrderProductType.php]: Done.

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php [hunk 10]
```diff
@@ -122,16 +125,21 @@ public function handle(GetOrderForViewing $query): OrderForViewing
...
         $order = $this->getOrder($query->getOrderId());
 
-        $taxMethod = $order->getTaxCalculationMethod() == PS_TAX_EXC ?
-            $this->translator->trans('Tax excluded', [], 'Admin.Global') :
-            $this->translator->trans('Tax included', [], 'Admin.Global');
+        $isTaxIncluded = $order->getTaxCalculationMethod() == PS_TAX_INC;
+
+        $taxMethod = $isTaxIncluded ?
+            $this->translator->trans('Tax included', [], 'Admin.Global') :
+            $this->translator->trans('Tax excluded', [], 'Admin.Global');
```
Comment (LouiseBonnard) [inline src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:132]: FYI, the order has changed compared to the removed lines. It now starts with _Tax included_ instead of _Tax excluded_. Perhaps you did it on purpose but in case you don't, I just wanted to notice it to you. ;-)

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/add_product_modal.html.twig [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+ * International Registered Trademark & Property of PrestaShop SA
+ *#}
+
+<div class="modal fade" id="addOrderProductModal" tabindex="-1" role="dialog">
+  {{ form_start(addProductToOrderForm, {'action': '#'}) }}
+  <div class="modal-dialog" role="document">
+    <div class="modal-content">
+      <div class="modal-header">
+        <h5 class="modal-title">
+          {{ 'Add a product'|trans({}, 'Admin.Catalog.Feature') }}
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/add_product_modal.html.twig]:
```suggestion
          {{ 'Add a product'|trans({}, 'Admin.Orderscustomers.Feature') }}
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 4 above)
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: > 🤦‍♂ I have an excuse, it will be needed for next PRs anyways. 😄

😆

File: src/PrestaShopBundle/Form/Admin/Sell/Order/UpdateOrderProductType.php (hunk 8 above)
Comment (matks) [reply src/PrestaShopBundle/Form/Admin/Sell/Order/UpdateOrderProductType.php]: Works too 👍 I wanted to make sure the relationship between Order and Product was obvious

**Comment by matks** [comment]
Only https://github.com/PrestaShop/PrestaShop/pull/15901#discussion_r333582669 missing, then code is valid 👍

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/add_product_modal.html.twig [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+              {{ form_widget(addProductToOrderForm.combination_id) }}
+            </div>
+          </div>
+        </div>
+
+        <div class="row">
+          <div class="col">
+            <div class="form-group">
+              <label class="form-control-label" for="{{ addProductToOrderForm.price_tax_excl.vars.id }}">
+                {{ 'Price tax excluded'|trans({}, 'Admin.Global') }}
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/add_product_modal.html.twig]:
```suggestion
                {{ 'Price tax excluded'|trans({}, 'Admin.Advparameters.Feature') }}
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/add_product_modal.html.twig [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+                <div class="input-group-append">
+                  <span class="input-group-text">{{ orderForViewing.currencyId }}</span>
+                </div>
+              </div>
+            </div>
+          </div>
+          <div class="col">
+            <div class="form-group mb-0">
+              <label class="form-control-label" for="{{ addProductToOrderForm.price_tax_incl.vars.id }}">
+                {{ 'Price tax included'|trans({}, 'Admin.Global') }}
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/add_product_modal.html.twig]:
```suggestion
                {{ 'Price tax included'|trans({}, 'Admin.Advparameters.Feature') }}
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/update_product_modal.html.twig [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        <button type="button" class="close" data-dismiss="modal" aria-label="{{ 'Close'|trans({}, 'Admin.Actions') }}">
+          <span aria-hidden="true">×</span>
+        </button>
+      </div>
+      <div class="modal-body">
+        <div class="row">
+          <div class="col">
+            <div class="form-group">
+              <label class="form-control-label" for="{{ updateOrderProductForm.price_tax_excl.vars.id }}">
+                {{ 'Price tax excluded'|trans({}, 'Admin.Global') }}
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/update_product_modal.html.twig]:
```suggestion
                {{ 'Price tax excluded'|trans({}, 'Admin.Advparameters.Feature') }}
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/update_product_modal.html.twig [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+                <div class="input-group-append">
+                  <span class="input-group-text">{{ orderForViewing.currencyId }}</span>
+                </div>
+              </div>
+            </div>
+          </div>
+          <div class="col">
+            <div class="form-group mb-0">
+              <label class="form-control-label" for="{{ updateOrderProductForm.price_tax_incl.vars.id }}">
+                {{ 'Price tax included'|trans({}, 'Admin.Global') }}
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/Modal/update_product_modal.html.twig]:
```suggestion
                {{ 'Price tax included'|trans({}, 'Admin.Advparameters.Feature') }}
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig [hunk 16]
```diff
@@ -74,60 +76,138 @@
...
-        </button>
+        {% if not orderForViewing.delivered %}
+          <button type="button" class="btn btn-outline-secondary" data-toggle="modal" data-target="#addOrderProductModal">
+            <i class="material-icons">add_circle_outline</i>
+            {{ 'Add a product'|trans({}, 'Admin.Orderscustomers.Feature') }}
+          </button>
+        {% endif %}
         <button type="button" class="btn btn-outline-secondary" data-toggle="modal" data-target="#addOrderDiscountModal">
           <i class="material-icons">confirmation_number</i>
           {{ 'Add a new discount'|trans({}, 'Admin.Orderscustomers.Feature') }}
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: Haven't we said ([here](https://github.com/PrestaShop/PrestaShop/pull/15388)) it should be _Add new cart rule_ (_Admin.Catalog.Feature_) instead?

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig [hunk 17]
```diff
@@ -74,60 +76,138 @@
...
         </button>
       </div>
     </div>
 
+    {% if orderForViewing.discounts.discounts is not empty %}
+      <table class="table">
+        <thead>
+          <tr>
+            <th>{{ 'Discount name'|trans({}, 'Admin.Global') }}</th>
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: I only have _Name_ (_Admin.Global_) in the mockups. cf. https://projects.invisionapp.com/d/main?origin=v7#/console/18147703/381874231/preview

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig [hunk 18]
```diff
@@ -74,60 +76,138 @@
...
-            </div>
-            <div class="col-sm text-center">
-              <p class="text-muted mb-0"><strong>{{ 'Shipping'|trans({}, 'Admin.Catalog.Feature') }}</strong></p>
-              <strong>99.99</strong>
+              <strong>{{ orderForViewing.prices.productsPriceFormatted }}</strong>
             </div>
+
+            {% if orderForViewing.prices.discountsAmountRaw %}
+              <div class="col-sm text-center">
+                <p class="text-muted mb-0"><strong>{{ 'Discounts'|trans({}, 'Admin.Orderscustomers.Feature') }}</strong></p>
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]:
```suggestion
                <p class="text-muted mb-0"><strong>{{ 'Discounts'|trans({}, 'Admin.Global') }}</strong></p>
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig [hunk 19]
```diff
@@ -74,60 +76,138 @@
...
+            {% if orderForViewing.prices.wrappingPriceRaw %}
+              <div class="col-sm text-center">
+                <p class="text-muted mb-0"><strong>{{ 'Wrapping'|trans({}, 'Admin.Orderscustomers.Feature') }}</strong></p>
+                <strong>{{ orderForViewing.prices.wrappingPriceFormatted }}</strong>
+              </div>
+            {% endif %}
+
+            {% if orderForViewing.prices.shippingPriceRaw %}
+              <div class="col-sm text-center">
+                <p class="text-muted mb-0"><strong>{{ 'Shipping'|trans({}, 'Admin.Catalog.Feature') }}</strong></p>
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig:196]: cf. https://github.com/PrestaShop/PrestaShop/pull/15388#discussion_r329576339

Comment (LouiseBonnard) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig:196]: Are we talking of the same thing here?

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php [hunk 20]
```diff
@@ -370,6 +433,7 @@ public function addPaymentAction(int $orderId, Request $request): RedirectRespon
     private function getErrorMessages(OrderException $e)
     {
         return [
+            CannotEditDeliveredOrderProductException::class => $this->trans('Cannot edit product for delivered order', 'Admin.Orderscustomers.Notification'),
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: I thought exceptions don't need translation domains, do they?

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig (hunk 17 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: That's right, I think I have followed legacy wording and I think it makes sense, otherwise, it requires user to what "name" is for, it's not explicitly marked that these are discounts. Wdyt? :)

File: src/PrestaShopBundle/Form/Admin/Sell/Order/UpdateOrderProductType.php (hunk 8 above)
Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Sell/Order/UpdateOrderProductType.php]: Done.

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php (hunk 10 above)
Comment (sarjon) [reply src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:132]: What exactly do you mean? :thinking:

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig (hunk 16 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: I can change, but "Discount" word is used a lot as well.

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 20 above)
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: It maps exception (logic error which is not translatable) to user understandable error which has to be translatable.

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig (hunk 19 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig:196]: Nope, it's different, it's order summary line called "Shipping" that displays shipping price.

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 20 above)
Comment (LouiseBonnard) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: @MatShir

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php (hunk 10 above)
Comment (LouiseBonnard) [reply src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:132]: Nothing important I guess, just look at the diff and you'll understand. :-)

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 20 above)
Comment (MatShir) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: @sarjon, I haven't specified on the specs but the old behaviours when the order was considered as delivered. The edit and delete button were hidden. Would be possible to keep the same behaviours?

Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: Yes, it is that way, but we still need to have error message just in case. :)

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php (hunk 10 above)
Comment (sarjon) [reply src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:132]: Sorry, but I can't see diff here. :confused:

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig (hunk 16 above)
Comment (LouiseBonnard) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: Indeed but the change of wording has already been merged [here](https://github.com/PrestaShop/PrestaShop/pull/15761/files), I guess it means we cannot change it again.

Sorry @sarjon, I try to gather all feedback to avoid confusion, it's quite difficult to follow this page's migration, they are many pull requests to double-check.
😕

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig (hunk 17 above)
Comment (LouiseBonnard) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: I agree but it might look weird to talk about cart rule before and discount after, doesn't it?

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 20 above)
Comment (LouiseBonnard) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: I suggest _You cannot edit the cart once the order delivered_ in _Admin.Orderscustomers.Notification_.

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php (hunk 10 above)
Comment (LouiseBonnard) [reply src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:132]: Nevermind!

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig (hunk 16 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: We can change, it's not a problem and it wasn't released yet. :)

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
Code is 👍 
Once we get validation for wording, we can move on QA 😄

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig (hunk 17 above)
Comment (sarjon) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: Yup, it might, but this wording is mixed a lot. Even since 1.7 side navigation was renamed from "Cart rules" (as it was in 1.6) to "Discounts" (here https://prnt.sc/pje0gk). 

So how do we deal with it? :thinking:

Comment (LouiseBonnard) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: In the back office, 'Discounts' is a generic word that refers to both cart rules and catalog price rules. If the first word is very clear for all users, it is quite uneasy to get at first sight what designate cart rules and catalog price rules. If the discount here only regards the possibility to add a cart rule, I suggest to call it cart rule:

- it is precise enough and familiarizes users to our discount terminology 
- it will also help users know what kind of discount to apply here

So, to summarize, I've just talked to @MatShir and here is what we recommend: using the button '_Add a discount_' (in _Admin.Orderscustomers.Feature_) on the Order Details page, then calling the pop-up frame '_Add new cart rule_' (in _Admin.Catalog.Feature_) and keeping the first field as '_Name_' (in _Admin.Global_).

Comment (sarjon) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/products.html.twig]: Done.

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php (hunk 20 above)
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: Done.

**Comment by sarjon** [comment]
Wording updated! :)

**Review by matks — DISMISSED** [review]

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
Merged, QA and product review will happen on next PR to be built on top of this one
